(#600) Notification system for settings (for crash logs and apk updates)

K1rakishou opened the pull request:
Closes #600

~~Not ready yet.~~

So, basically, instead of spamming AlertDialogs on every important update (like new apk update or new crash log that awaits to be either uploaded or not) there is now a more user-friendly system which does not nag the user with any windows whatsoever. Instead, a small alert icon is being shown in specific setting option (and in the drawer too) so the user sees it every time he opens the drawer or the settings. Once the conditions for this icon are all satisfied it disappears (like all crash logs are either uploaded or deleted). If there are multiple notifications (say apk update and couple crash logs) then a notification with highest priority is shown in the drawer (once it's cleared the other one replaces it, etc).

It's pretty easy to add new notifications, too.

Oh yeah, you can now also review crash logs and decide which you upload and which not (you can even delete them all). No crash logs are now being uploaded automatically.

K1rakishou committed:
(#600) Basic implementation of settings notification system

- Implement SettingsNotificationManager to accumulate update events and distribute them to observers.
- Implement displaying of the notifications in the settings' views.

K1rakishou committed:
(#600) Crash log notifications work!

K1rakishou committed:
(#600) Remove redundant TODO

K1rakishou committed:
(#600) Crash log notifications work!

K1rakishou committed:
(#600) "New apk update" notifications work!

K1rakishou commented:
Alright, now it's ready.

K1rakishou committed:
(#600) Merge with dev + color fixes

K1rakishou commented:
I had to change a couple of things in UpdateManager so you better review it more carefully.

K1rakishou committed:
(#600) autoCrashLogsUpload -> collectCrashLogs, delete all crash logs when disabling crash log collect setting

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/settings/MainSettingsController.java:
```diff
@@ -204,4 +251,73 @@ private void setupAboutGroup() {
+        openReportProblemController();
+    }
+
+    private void suggestSendingCrashLogs(int crashLogsCount) {
```
Probably should be called "SuggestReviewCrashLogs"

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/layout/crashlogs/CrashLog.kt:
```diff
@@ -0,0 +1,24 @@
+
+import java.io.File
+
+data class CrashLog(val file: File, val fileName: String, var send: Boolean) {
```
"send" should actually be called "selected" or something

K1rakishou committed:
(#600) Code review fixes

K1rakishou committed:
(#600) Merge branch 'cheese-multi-feature' into (#600)-notification-system-for-settings

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/manager/UpdateManager.java

K1rakishou committed:
(#600) Merge branch 'cheese-multi-feature' into (#600)-notification-system-for-settings

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/Chan.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/manager/UpdateManager.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/settings/ChanSettings.java
#	Kuroba/app/src/main/res/values/strings.xml

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/manager/SettingsNotificationManager.kt:
```diff
@@ -0,0 +1,94 @@
+import io.reactivex.processors.BehaviorProcessor
+
+class SettingsNotificationManager {
+    private val notifications: MutableSet<SettingNotificationType> = mutableSetOf()
```
Probably should add "@GuardedBy("this")"

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/manager/SettingsNotificationManager.kt:
```diff
@@ -0,0 +1,94 @@
+     * */
+    fun listenForNotificationUpdates(): Flowable<Unit> = activeNotificationsSubject
+            .onBackpressureBuffer()
+            .observeOn(AndroidSchedulers.mainThread())
```
Use `hide()` here.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/manager/UpdateManager.java:
```diff
@@ -168,7 +179,7 @@ private void runUpdateApi(final boolean manual) {
             //region Release build
             volleyRequestQueue.add(new UpdateApiRequest(response -> {
-                if (!processUpdateApiResponse(response) && manual && BackgroundUtils.isInForeground()) {
+                if (!processUpdateApiResponse(response, manual && BackgroundUtils.isInForeground()) && manual) {
```
Uhh, something broke here after merge? This doesn't seem correct.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/manager/UpdateManager.java:
```diff
@@ -227,25 +240,49 @@ private void runUpdateApi(final boolean manual) {
     }
 
+    private void notifyNewApkUpdate() {
+        ChanSettings.hasNewApkUpdate.set(true);
```
This is not a setting. This is a persistable app state. I think I should extract it into a separate class that uses it's own shared preferences file.

Adamantcheese replied:
You probably won't need a setting at all since the update API is run on every startup.

K1rakishou replied:
But it only checks for the new apk only once in 24h, or is it not?

Adamantcheese replied:
It checks for a new APK only when the app is started cold, with a minimum delay of 24h for dev and 5 days for stable. There's no scheduled task to check for an update, just like how it was in Clover.

K1rakishou replied:
Well that's why I needed "hasNewApkUpdate" flag. Because if we start the app during the 24h or 5d window when we do not check for the updates, and there were actually an update the last time we checked it, the notification won't show up.

Adamantcheese replied:
Ah okay, I understand. You're setting it in case someone runs the updater manually, but defers installation to a later point.

Adamantcheese replied:
It's still fine to keep it in ChanSettings anyways, there's a bunch of things in there that aren't strictly settings.

K1rakishou replied:
Well those will have to be moved too. But we can do that later. Or I could move them.

Adamantcheese replied:
Please just keep it to one thing per PR if possible, these super huge PRs with all sorts of barely related stuff in them makes it difficult to review.

K1rakishou replied:
Alright then I will just move this one setting to a new class.

K1rakishou replied:
Done.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/DrawerController.java:
```diff
@@ -62,14 +68,20 @@
     protected RecyclerView recyclerView;
     protected DrawerAdapter drawerAdapter;
+    @Nullable
+    private CompositeDisposable compositeDisposable = null;
```
Probably shouldn't make it nullable. Just create it once and then use clear() instead of dispose()

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/settings/SettingsController.java:
```diff
@@ -39,30 +44,40 @@
     // Very user unfriendly.
     protected List<SettingView> requiresRestart = new ArrayList<>();
-
+    protected CompositeDisposable compositeDisposable;
```
Don't make it nullable. Create it once and then use clear() instead of dispose()

K1rakishou commented:
~~I should also only upload crashes that are not older than, let's say, 3 days (and delete older crashes) to avoid endless spam. Also should check and handle http codes 429 (when client attempted to send way too many crash logs) and 500.~~

Already done in another PR and already merged.

K1rakishou committed:
(#600) Code review fixes

K1rakishou committed:
(#600) Merge branch 'cheese-multi-feature' into (#600)-notification-system-for-settings

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/Chan.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/settings/DeveloperSettingsController.java

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/manager/ReportManager.kt:
```diff
@@ -169,37 +179,97 @@ class ReportManager(
+        }
+
+        potentialCrashLogs.asSequence()
+                .filter { file -> file.name.startsWith(CRASH_LOG_FILE_NAME_PREFIX) }
```
Probably should remove this `filter` operator because only crash logs are supposed to be stored in that directory anyway so everything should be deleted.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/manager/SettingsNotificationManager.kt:
```diff
@@ -0,0 +1,97 @@
+    private val notifications: MutableSet<SettingNotificationType> = mutableSetOf()
+
+    /**
+     * A reactive stream that is being used to notify observer about [notifications] changes
```
observers

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/manager/SettingsNotificationManager.kt:
```diff
@@ -0,0 +1,97 @@
+            return SettingNotificationType.ApkUpdate
+        }
+
+        if (contains(SettingNotificationType.CrashLogs)) {
```
Maybe I should rename it to CrashLog?

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/settings/ChanState.kt:
```diff
@@ -0,0 +1,24 @@
+import com.github.adamantcheese.chan.utils.Logger
+
+
+object ChanState {
```
Actually this class is still unmockable. I need to create an interface with all the settings methods and inherit this singleton from the interface. And with that I will be able to create mocks of the interface.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/settings/ChanState.kt:
```diff
@@ -0,0 +1,24 @@
+
+    // Why? So it can be mocked in tests.
+    @JvmStatic
+    open fun hasNewApkUpdate() = hasNewApkUpdate
```
And I guess I should provide two methods: get and set so it doesn't need to return String/Boolean/whatever setting which I will have to mock too.

K1rakishou committed:
(#600) Coder review fixes

K1rakishou committed:
(#600) Merge branch 'cheese-multi-feature' into (#600)-notification-system-for-settings

# Conflicts:
#	Kuroba/app/src/main/res/values/strings.xml

K1rakishou committed:
(#600) Merge branch 'cheese-multi-feature' into (#600)-notification-system-for-settings

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/layout/ReportProblemLayout.kt
